# Data Wrangling

Data and information retrieved from the UCI Machine Learning Repository. <br>
https://archive.ics.uci.edu/ml/datasets/Drug+consumption+%28quantified%29 <br>
https://www.healthline.com/health/big-five-personality-traits#takeaway

### Data Collection

In [1]:
import pandas as pd
import os
import numpy as np

In [2]:
print(os.getcwd())
print(os.listdir())

C:\Users\filia\Springboard DS\Capstone 2\Personality Type Drug Consumption Risk
['.ipynb_checkpoints', 'drug_consumption.data', 'Untitled.ipynb']


In [3]:
#data does not contain column names
df = pd.read_csv('drug_consumption.data', header=None)
df.head()

0        1        2        3        4        5        6        7        8   \
0   1  0.49788  0.48246 -0.05921  0.96082  0.12600  0.31287 -0.57545 -0.58331   
1   2 -0.07854 -0.48246  1.98437  0.96082 -0.31685 -0.67825  1.93886  1.43533   
2   3  0.49788 -0.48246 -0.05921  0.96082 -0.31685 -0.46725  0.80523 -0.84732   
3   4 -0.95197  0.48246  1.16365  0.96082 -0.31685 -0.14882 -0.80615 -0.01928   
4   5  0.49788  0.48246  1.98437  0.96082 -0.31685  0.73545 -1.63340 -0.45174   

        9   ...   22   23   24   25   26   27   28   29   30   31  
0 -0.91699  ...  CL0  CL0  CL0  CL0  CL0  CL0  CL0  CL2  CL0  CL0  
1  0.76096  ...  CL4  CL0  CL2  CL0  CL2  CL3  CL0  CL4  CL0  CL0  
2 -1.62090  ...  CL0  CL0  CL0  CL0  CL0  CL0  CL1  CL0  CL0  CL0  
3  0.59042  ...  CL0  CL0  CL2  CL0  CL0  CL0  CL0  CL2  CL0  CL0  
4 -0.30172  ...  CL1  CL0  CL0  CL1  CL0  CL0  CL2  CL2  CL0  CL0  

[5 rows x 32 columns]

In [4]:
#assign column names to data
df.columns= ['ID','Age','Gender','Education','Country','Ethnicity','Neuroticism','Extraversion','Openness_to_Experience','Agreeableness','Conscientiousness','Impulsiveness','Sensation-Seeking','Alcohol','Amphetamines','Amyl','Benzos','Caffeine','Cannabis','Chocolate','Cocaine','Crack','Ecstasy','Heroin','Ketamine','Legal Highs','LSD','Methadone','Mushrooms','Nicotine','Semer','VSA']
df.tail()

ID      Age   Gender  Education  Country  Ethnicity  Neuroticism  \
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685     -1.19430   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685     -0.24649   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685      1.13281   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685      0.91093   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685     -0.46725   

      Extraversion  Openness_to_Experience  Agreeableness  ...  Ecstasy  \
1880       1.74091                 1.88511        0.76096  ...      CL0   
1881       1.74091                 0.58331        0.76096  ...      CL2   
1882      -1.37639                -1.27553       -1.77200  ...      CL4   
1883      -1.92173                 0.29338       -1.62090  ...      CL3   
1884       2.12700                 1.65653        1.11406  ...      CL3   

      Heroin  Ketamine Legal Highs  LSD Methadone Mushrooms Nicotine Semer  \
1880     CL0       CL0         CL3  CL3       CL0       CL0      CL0   CL0   
1881     CL0       CL0         CL3  CL5       CL4       CL4      CL5   CL0   
1882     CL0       CL2         CL0  CL2       CL0       CL2      CL6   CL0   
1883     CL0       CL0         CL3  CL3       CL0       CL3      CL4   CL0   
1884     CL0       CL0         CL3  CL3       CL0       CL3      CL6   CL0   

      VSA  
1880  CL5  
1881  CL0  
1882  CL0  
1883  CL0  
1884  CL2  

[5 rows x 32 columns]

### Data Definition

1. __ID:__ used for reference <br> <br>

2. __Age:__ age of participant and has one of the values
Value Meaning Cases Fraction <br>
-0.95197 18-24 643 34.11% <br>
-0.07854 25-34 481 25.52% <br>
0.49788 35-44 356 18.89% <br>
1.09449 45-54 294 15.60% <br>
1.82213 55-64 93 4.93% <br>
2.59171 65+ 18 0.95% <br> <br>

3. __Gender:__ gender of participant <br>
Value Meaning Cases Fraction <br>
0.48246 Female 942 49.97% <br>
-0.48246 Male 943 50.03% <br> <br>

4. __Education:__ level of education of participant <br>
Value Meaning Cases Fraction <br>
-2.43591 Left school before 16 years 28 1.49% <br>
-1.73790 Left school at 16 years 99 5.25% <br>
-1.43719 Left school at 17 years 30 1.59% <br>
-1.22751 Left school at 18 years 100 5.31% <br>
-0.61113 Some college or university, no certificate or degree 506 26.84% <br>
-0.05921 Professional certificate/ diploma 270 14.32% <br>
0.45468 University degree 480 25.46% <br>
1.16365 Masters degree 283 15.01% <br>
1.98437 Doctorate degree 89 4.72% <br> <br>

5. __Country:__ country of current residence of participant <br>
Value Meaning Cases Fraction <br>
-0.09765 Australia 54 2.86% <br>
0.24923 Canada 87 4.62% <br>
-0.46841 New Zealand 5 0.27% <br>
-0.28519 Other 118 6.26% <br>
0.21128 Republic of Ireland 20 1.06% <br>
0.96082 UK 1044 55.38% <br>
-0.57009 USA 557 29.55% <br> <br>

6. __Ethnicity:__ ethnicity of participant <br>
Value Meaning Cases Fraction <br>
-0.50212 Asian 26 1.38% <br>
-1.10702 Black 33 1.75% <br>
1.90725 Mixed-Black/Asian 3 0.16% <br>
0.12600 Mixed-White/Asian 20 1.06% <br>
-0.22166 Mixed-White/Black 20 1.06% <br>
0.11440 Other 63 3.34% <br>
-0.31685 White 1720 91.25% <br> <br>

7. __Neuroticism:__ Nscore is NEO-FFI-R Neuroticism <br>
Nscore Cases Value Nscore Cases Value Nscore Cases Value <br>
12 1 -3.46436 29 60 -0.67825 46 67 1.02119 <br>
13 1 -3.15735 30 61 -0.58016 47 27 1.13281 <br>
14 7 -2.75696 31 87 -0.46725 48 49 1.23461 <br>
15 4 -2.52197 32 78 -0.34799 49 40 1.37297 <br>
16 3 -2.42317 33 68 -0.24649 50 24 1.49158 <br>
17 4 -2.34360 34 76 -0.14882 51 27 1.60383 <br>
18 10 -2.21844 35 69 -0.05188 52 17 1.72012 <br>
19 16 -2.05048 36 73 0.04257 53 20 1.83990 <br>
20 24 -1.86962 37 67 0.13606 54 15 1.98437 <br>
21 31 -1.69163 38 63 0.22393 55 11 2.12700 <br>
22 26 -1.55078 39 66 0.31287 56 10 2.28554 <br>
23 29 -1.43907 40 80 0.41667 57 6 2.46262 <br>
24 35 -1.32828 41 61 0.52135 58 3 2.61139 <br>
25 56 -1.19430 42 77 0.62967 59 5 2.82196 <br>
26 57 -1.05308 43 49 0.73545 60 2 3.27393 <br>
27 65 -0.92104 44 51 0.82562 <br>
28 70 -0.79151 45 37 0.91093 <br> <br>

8. __Extraversion:__ Escore is NEO-FFI-R Extraversion <br>
Escore Cases Value Escore Cases Value Escore Cases Value <br>
16 2 -3.27393 31 55 -1.23177 45 91 0.80523 <br>
18 1 -3.00537 32 52 -1.09207 46 69 0.96248 <br>
19 6 -2.72827 33 77 -0.94779 47 64 1.11406 <br>
20 3 -2.53830 34 68 -0.80615 48 62 1.28610 <br>
21 3 -2.44904 35 58 -0.69509 49 37 1.45421 <br>
22 8 -2.32338 36 89 -0.57545 50 25 1.58487 <br>
23 5 -2.21069 37 90 -0.43999 51 34 1.74091 <br>
24 9 -2.11437 38 106 -0.30033 52 21 1.93886 <br>
25 4 -2.03972 39 107 -0.15487 53 15 2.12700 <br>
26 21 -1.92173 40 130 0.00332 54 10 2.32338 <br>
27 23 -1.76250 41 116 0.16767 55 9 2.57309 <br>
28 23 -1.63340 42 109 0.32197 56 2 2.85950 <br>
29 32 -1.50796 43 105 0.47617 58 1 3.00537 <br>
30 38 -1.37639 44 103 0.63779 59 2 3.27393 <br> <br>

9. __Openness to Experience:__ Oscore is NEO-FFI-R Openness to experience <br>
Oscore Cases Value Oscore Cases Value Oscore Cases Value <br>
24 2 -3.27393 38 64 -1.11902 50 83 0.58331 <br>
26 4 -2.85950 39 60 -0.97631 51 87 0.72330 <br>
28 4 -2.63199 40 68 -0.84732 52 87 0.88309 <br>
29 11 -2.39883 41 76 -0.71727 53 81 1.06238 <br>
30 9 -2.21069 42 87 -0.58331 54 57 1.24033 <br>
31 9 -2.09015 43 86 -0.45174 55 63 1.43533 <br>
32 13 -1.97495 44 101 -0.31776 56 38 1.65653 <br>
33 23 -1.82919 45 103 -0.17779 57 34 1.88511 <br>
34 25 -1.68062 46 134 -0.01928 58 19 2.15324 <br>
35 26 -1.55521 47 107 0.14143 59 13 2.44904 <br>
36 39 -1.42424 48 116 0.29338 60 7 2.90161 <br>
37 51 -1.27553 49 98 0.44585 <br> <br>

10. __Agreeableness:__ Ascore is NEO-FFI-R Agreeableness <br>
Ascore Cases Value Ascore Cases Value Ascore Cases Value <br>
12 1 -3.46436 34 42 -1.34289 48 104 0.76096 <br>
16 1 -3.15735 35 45 -1.21213 49 85 0.94156 <br>
18 1 -3.00537 36 62 -1.07533 50 68 1.11406 <br>
23 1 -2.90161 37 83 -0.91699 51 58 1.2861 <br>
24 2 -2.78793 38 82 -0.76096 52 39 1.45039 <br>
25 1 -2.70172 39 102 -0.60633 53 36 1.61108 <br>
26 7 -2.53830 40 98 -0.45321 54 36 1.81866 <br>
27 7 -2.35413 41 114 -0.30172 55 16 2.03972 <br>
28 8 -2.21844 42 101 -0.15487 56 14 2.23427 <br>
29 13 -2.07848 43 105 -0.01729 57 8 2.46262 <br>
30 18 -1.92595 44 118 0.13136 58 7 2.75696 <br>
31 24 -1.77200 45 112 0.28783 59 1 3.15735 <br>
32 30 -1.62090 46 100 0.43852 60 1 3.46436 <br>
33 34 -1.47955 47 100 0.59042 <br> <br>

11. __Conscientiousness:__ Cscore is NEO-FFI-R Conscientiousness <br>
Cscore Cases Value Cscore Cases Value Cscore Cases Value <br>
17 1 -3.46436 32 39 -1.25773 46 113 0.58489 <br>
19 1 -3.15735 33 49 -1.13788 47 95 0.7583 <br>
20 3 -2.90161 34 55 -1.01450 48 95 0.93949 <br>
21 2 -2.72827 35 55 -0.89891 49 76 1.13407 <br>
22 5 -2.57309 36 69 -0.78155 50 47 1.30612 <br>
23 5 -2.42317 37 81 -0.65253 51 43 1.46191 <br>
24 6 -2.30408 38 77 -0.52745 52 34 1.63088 <br>
25 9 -2.18109 39 87 -0.40581 53 28 1.81175 <br>
26 13 -2.04506 40 97 -0.27607 54 27 2.04506 <br>
27 13 -1.92173 41 99 -0.14277 55 13 2.33337 <br>
28 25 -1.78169 42 105 -0.00665 56 8 2.63199 <br>
29 24 -1.64101 43 90 0.12331 57 3 3.00537 <br>
30 29 -1.51840 44 111 0.25953 59 1 3.46436 <br>
31 41 -1.38502 45 111 0.41594 <br> <br>

12. __Impulsiveness:__ Impulsiveness measured by BIS-11 <br>
_*Items are aggregated, and the higher BIS-11 scores, the higher
the impulsivity level_ <br>
Impulsiveness Cases Fraction <br>
-2.55524 20 1.06% <br>
-1.37983 276 14.64% <br>
-0.71126 307 16.29% <br>
-0.21712 355 18.83% <br>
0.19268 257 13.63% <br>
0.52975 216 11.46% <br>
0.88113 195 10.34% <br>
1.29221 148 7.85% <br>
1.86203 104 5.52% <br>
2.90161 7 0.37% <br> <br>

13. __Sensation-Seeking:__ SS is sensation seeking measured by ImpSS <br>
_*Although the ImpSS combines the traits of impulsivity and sensation-seeking,
it is regarded as a measure of a general sensation-seeking trait. Assumption
is the higher the ss score, the higher degree of sensation seeking_ <br>
SS Cases Fraction <br>
-2.07848 71 3.77% <br>
-1.54858 87 4.62% <br>
-1.18084 132 7.00% <br>
-0.84637 169 8.97% <br>
-0.52593 211 11.19% <br>
-0.21575 223 11.83% <br>
0.07987 219 11.62% <br>
0.40148 249 13.21% <br>
0.76540 211 11.19% <br>
1.22470 210 11.14% <br>
1.92173 103 5.46% <br> <br>

14. __Alcohol:__ alcohol consumption <br>

15. __Amphetamines:__ amphetamines consumption <br>

16. __Amyl__ amyl nitrite consumption <br>

17. __Benzos:__ benzodiazepine consumption <br>

18. __Caffeine:__ caffeine consumption <br>

19. __Cannabis:__ cannabis consumption <br>

20. __Chocolate:__ chocolate consumption <br>

21. __Cocaine:__ cocaine consumption <br>

22. __Crack:__ crack consumption <br>

23. __Ecstasy:__ ecstasy consumption <br>

24. __Heroin:__ heroin consumption <br>

25. __Ketamine:__ ketamine consumption <br>

26. __Legal Highs:__ legal highs consumption <br>

27. __LSD:__ lysergic acid diethylamide consumption <br>

28. __Methadone:__ methadone consumption <br>

29. __Mushrooms:__ magic mushrooms consumption <br>

30. __Nicotine:__ nicotine consumption <br>

31. __Semer:__ fictitious drug Semeron consumption <br>

32. __VSA:__ volatile substance abuse consumption <br> <br>

__*All values for drug consumption are classified by the following codes:__ <br>
CL0 - Never Used <br>
CL1 - Used over a Decade Ago <br>
CL2 - Used in Last Decade <br>
CL3 - Used in Last Year <br>
CL4 - Used in Last Month <br>
CL5 - Used in Last Week <br>
CL6 - Used in Last Day

### Make Data Values Meaningful

In [5]:
age_range = ['18-24','25-34','35-44','45-54','55-64','65+']
age_vals = sorted(df.Age.unique())
#print(age_vals)
df['Age'] = df.Age.replace(age_vals, age_range)

In [6]:
gender = ['Male','Female']
gender_vals = sorted(df.Gender.unique())
#print(gender_vals)
df['Gender'] = df.Gender.replace(gender_vals, gender)

In [7]:
edu = ['Left school before 16','Left school at 16','Left school at 17','Left school at 18','Some college/university','Professional certificate/diploma','Bachelors','Masters','Doctorate']
edu_vals = sorted(df.Education.unique())
#print(edu_vals)
df['Education'] = df.Education.replace(edu_vals, edu)

In [8]:
cntry = ['Australia','Canada','New Zealand','Other','Republic of Ireland','UK','USA']
cntry_vals = [-0.09765,0.24923,-0.46841000000000005,-0.28519,0.21128000000000002,0.9608200000000001,-0.57009]

df['Country'] = df.Country.replace(cntry_vals, cntry)

In [9]:
ethn = ['Asian','Black','Mixed-Black/Asian','Mixed-White/Asian','Mixed-White/Black','Other','White']
ethn_vals = [-0.50212,-1.1070200000000001,1.90725,0.12600,-0.22166,0.11440,-0.31685]

df['Ethnicity'] = df.Ethnicity.replace(ethn_vals, ethn)

In [10]:
nscore = np.arange(12,61)
nscore_vals = sorted(df.Neuroticism.unique())
#print(nscore_vals)
df['Neuroticism'] = df.Neuroticism.replace(nscore_vals, nscore)

In [11]:
escore = [x for x in range(16,60) if x not in [17,57]]
escore_vals = sorted(df.Extraversion.unique())
#print(escore)
#print(escore_vals)
df['Extraversion'] = df.Extraversion.replace(escore_vals, escore)

In [12]:
oscore = [x for x in range(24,61) if x not in [25,27]]
oscore_vals = sorted(df.Openness_to_Experience.unique())
#print(oscore)
#print(oscore_vals)
df['Openness_to_Experience'] = df.Openness_to_Experience.replace(oscore_vals, oscore)

In [13]:
ascore = [x for x in range(12,61) if x not in [13,14,15,17,19,20,21,22]]
ascore_vals = sorted(df.Agreeableness.unique())
#print(ascore)
#print(ascore_vals)
df['Agreeableness'] = df.Agreeableness.replace(ascore_vals, ascore)

In [14]:
cscore = [x for x in range(17,60) if x not in [18,58]]
cscore_vals = sorted(df.Conscientiousness.unique())
#print(cscore)
#print(cscore_vals)
df['Conscientiousness'] = df.Conscientiousness.replace(cscore_vals, cscore)

In [15]:
#drugs_df = df.iloc[:,13:32]
#print(drugs_df.columns)
#drugs_df.head()

In [16]:
drug_code = ['CL0', 'CL1','CL2','CL3','CL4','CL5','CL6']
use = ['Never used','Used over a decade ago','Used in last decade','Used in last year','Used in last month','Used in last week','Used in last day']

df.iloc[:,13:32] = df.iloc[:,13:32].replace(drug_code, use)
df.head()

ID    Age  Gender                         Education Country  \
0   1  35-44  Female  Professional certificate/diploma      UK   
1   2  25-34    Male                         Doctorate      UK   
2   3  35-44    Male  Professional certificate/diploma      UK   
3   4  18-24  Female                           Masters      UK   
4   5  35-44  Female                         Doctorate      UK   

           Ethnicity  Neuroticism  Extraversion  Openness_to_Experience  \
0  Mixed-White/Asian         39.0          36.0                    42.0   
1              White         29.0          52.0                    55.0   
2              White         31.0          45.0                    40.0   
3              White         34.0          34.0                    46.0   
4              White         43.0          28.0                    43.0   

   Agreeableness  ...                 Ecstasy      Heroin  \
0           37.0  ...              Never used  Never used   
1           48.0  ...      Used in last month  Never used   
2           32.0  ...              Never used  Never used   
3           47.0  ...              Never used  Never used   
4           41.0  ...  Used over a decade ago  Never used   

              Ketamine             Legal Highs                  LSD  \
0           Never used              Never used           Never used   
1  Used in last decade              Never used  Used in last decade   
2           Never used              Never used           Never used   
3  Used in last decade              Never used           Never used   
4           Never used  Used over a decade ago           Never used   

           Methadone               Mushrooms             Nicotine       Semer  \
0         Never used              Never used  Used in last decade  Never used   
1  Used in last year              Never used   Used in last month  Never used   
2         Never used  Used over a decade ago           Never used  Never used   
3         Never used              Never used  Used in last decade  Never used   
4         Never used     Used in last decade  Used in last decade  Never used   

          VSA  
0  Never used  
1  Never used  
2  Never used  
3  Never used  
4  Never used  

[5 rows x 32 columns]

In [17]:
#verify values have been replaced
#for col in df.columns:
#    print(df[col].unique())

### Dataset Exploration

In [18]:
df.shape

(1885, 32)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1885 entries, 0 to 1884
Data columns (total 32 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ID                      1885 non-null   int64  
 1   Age                     1885 non-null   object 
 2   Gender                  1885 non-null   object 
 3   Education               1885 non-null   object 
 4   Country                 1885 non-null   object 
 5   Ethnicity               1885 non-null   object 
 6   Neuroticism             1885 non-null   float64
 7   Extraversion            1885 non-null   float64
 8   Openness_to_Experience  1885 non-null   float64
 9   Agreeableness           1885 non-null   float64
 10  Conscientiousness       1885 non-null   float64
 11  Impulsiveness           1885 non-null   float64
 12  Sensation-Seeking       1885 non-null   float64
 13  Alcohol                 1885 non-null   object 
 14  Amphetamines            1885 non-null   

In [20]:
df.drop(columns='ID',axis=1).describe().T

count       mean       std       min       25%  \
Neuroticism             1885.0  35.921485  9.135869  12.00000  29.00000   
Extraversion            1885.0  39.576127  6.771769  16.00000  35.00000   
Openness_to_Experience  1885.0  45.762334  6.579641  24.00000  41.00000   
Agreeableness           1885.0  42.866313  6.438106  12.00000  39.00000   
Conscientiousness       1885.0  41.437135  6.966625  17.00000  37.00000   
Impulsiveness           1885.0   0.007216  0.954435  -2.55524  -0.71126   
Sensation-Seeking       1885.0  -0.003292  0.963701  -2.07848  -0.52593   

                             50%       75%       max  
Neuroticism             36.00000  42.00000  60.00000  
Extraversion            40.00000  44.00000  59.00000  
Openness_to_Experience  46.00000  51.00000  60.00000  
Agreeableness           43.00000  48.00000  60.00000  
Conscientiousness       42.00000  46.00000  59.00000  
Impulsiveness           -0.21712   0.52975   2.90161  
Sensation-Seeking        0.07987   0.76540   1.92173

### Check For Nulls and Duplicate Rows

In [21]:
#Function that calculates percentage of missing data per column
def missingData(dataframe):
    missing = False
    for item in (dataframe.isnull().sum()).iteritems():
        if item[1] > 0:
            print('Missing Data percentage for '+item[0]+' is {:2.2%}'.format((item[1]/dataframe.shape[0])) )
            missing = True
    if not missing:
        print('Found no missing values.')

missingData(df)

Found no missing values.


In [22]:
num_dups = len(df) - len(df.drop_duplicates())
print(num_dups)
df = df.drop_duplicates()

0


### Export Data to CSV for EDA

In [23]:
df.to_csv('drug_consumption.csv')
os.listdir()

['.ipynb_checkpoints',
 'drug_consumption.csv',
 'drug_consumption.data',
 'Untitled.ipynb']